keps/sig-cluster-lifecycle/wgs/2141-instance-specific-componentconfig/README.md

```diff
@@ -0,0 +1,616 @@
+func ConvertYAMLToJSON(yamlSharable, yamlInstance []byte) ([]byte, []byte, error) {
+    jsonSharable, err := yaml.YAMLToJSON(yamlSharable)
+    if err != nil {
+        return nil, nil, err
+    }
+
+    jsonInstance, err := yaml.YAMLToJSON(yamlInstance)
+    if err != nil {
+        return nil, nil, err
+    }
+
+    return jsonSharable, jsonInstance, nil
+}
+
+// PatchConfiguration applies a strategic merge patch and returns the patched JSON result,
+// it is necessary to convert both configurations from YAML to JSON before applying the merge.
+func PatchConfiguration(jsonSharable, jsonInstance []byte) []byte, error {
+    obj := &kubeletv1beta1.KubeletConfiguration{}
+
+    patched, err := strategicpatch.StrategicMergePatch(jsonSharable, jsonInstance, obj)
+    if err != nil {
+        return []byte{}, err
+    }
+
+    return patched, err
+}
+```
+
+The steps for this merges is simpler than the alternative approach:
+
+1. Convert `shared.yaml` configuration from YAML to JSON.
+2. Convert `instance.yaml` configuration from YAML to JSON.
+3. Patch the shared JSON configuration with the instance JSON, the
+   apimachinery strategic merge patch.
+
+In the diagram below it's possible to have a better visualization of the
+strategic merge flow:
+
+![component-smp-merge](images/component-smp-merge.png)
+
+### Limitations
+
+This approach has the following limitations. We may decide to address these in the
+future but it is not a goal of this KEP.
+
+* Users can't currently choose the merge strategy. It might make sense for users
+  to be able to choose, since they are the authors of the configuration.
+* This merge strategy requires both objects to have the same Kind. In the future
+  it may turn out to be a good idea to more carefully categorize fields as shared
+  or instance-specific.
+* This merge strategy requires both objects to have the same API version. In the
+  future, it may be useful to be able to convert versions for one or the other
+  prior to the merge, but set/unset-preserving conversions need to be possible first.
+* If a component has multiple configuration Kinds for sub-components (for instance,
+  some conversations around controller-manager's config have leaned in this
+  direction), it may be useful for the config and instance-config files to contain
+  multiple YAML documents, and merge objects paired by Kind between the two.
+  We can consider supporting this in future iterations.
+
+### Test Plan
+
+ComponentConfig APIs are typically tested similarly to API server APIs, for
+example by testing round-trip conversions. Some tests specifically exercise
+the ability to share a config object, such as the Dynamic Kubelet Config tests.
+
+Components that opt-in to instance-specific Kinds should extend their existing
+tests to include the instance-specific Kinds, except where those tests are
+explicitly testing the ability to share a config object across concurrent
+instances.
+
+### Graduation Criteria
```

Comment:
When this is targetted to a release, I'd expect to see more specific Graduation Criteria for example:
Alpha: X components using this approach
Beta: X beta component using this approach with some specific details etc..

(guessing here but you get the idea :smile: )

Reply:
Let's add implementing the --instance-config _flag_ in alpha for kubelet and kube-proxy, graduating based on feedback, and experience, and sufficient test coverage of those implementations (and considering a shared implementation in component-base if that makes sense).

Beta graduation criteria for the KEP can be both components have a stable, consistent behavior of --instance-config, and both have Beta ComponentConfig APIs (kubelet is beta already, kube-proxy is not, but since a goal of this KEP is to unblock kube-proxy, I think this is a good criteria). 

GA graduation of the KEP can be concurrent with both kubelet and kube-proxy having GA ComponentConfig APIs. I don't expect much else to change between beta and GA. 

Want to clarify, also, that independent components' configuration APIs must follow the typical API graduation process and any other criteria defined by the component. This KEP is intended to become the recommended way for components to provide a solution for instance-specific config, and proof that it works well enough for component APIs to graduate through their own process should be sufficient.